enterprise/tpa_pipeline.py

```diff
@@ -0,0 +1,153 @@
+"""
+Module provides elements to be used in third-party auth pipeline
+"""
+from django.core.urlresolvers import reverse
+from django.shortcuts import redirect
+from django.utils.translation import ugettext as _
+
+try:
+    from third_party_auth.provider import Registry
+except ImportError:
+    Registry = None
+
+from enterprise.models import EnterpriseCustomer, UserDataSharingConsentAudit
+from enterprise.utils import NotConnectedToEdX
+
+from social.pipeline import partial
+
+class MultipleCustomersFound(Exception):
+    """
+    Raise this exception if there are multiple customers tied to an IdP.
+    """
+
+
+def get_enterprise_customer_for_request(request):
+    """
+    Get the EnterpriseCustomer associated with a particular request.
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+    pipeline = request.session.get('partial_pipeline')
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_running_pipeline(pipeline):
+    """
+    Get the EnterpriseCustomer associated with a running pipeline
+    """
+    if Registry is None:
+        raise NotConnectedToEdX(
+            _("This package must be installed in an EdX environment to look up third-party auth providers.")
+        )
+
+    provider = Registry.get_from_pipeline(pipeline)
+    return get_enterprise_customer_for_sso(provider)
+
+
+def get_enterprise_customer_for_sso(provider):
+    """
+    Get the EnterpriseCustomer object tied to an identity provider.
+    """
+    if provider is None:
+        return None
+    try:
+        return EnterpriseCustomer.objects.get(identity_provider=provider.idp_slug)
+    except EnterpriseCustomer.MultipleObjectsReturned:
+        raise MultipleCustomersFound(
+            _("Multiple EnterpriseCustomer objects are tied to the same identity provider; expected one.")
```

Comment:
Why are we not enforcing this at the database level instead, by making the `identity_provider` field `unique=True` ?